File: backend/tasks/agents/analyst/onboarding_agents.py
"""OnboardingAgent - Fast, focused agents for onboarding steps."""

import json
import time
from tasks import register_agent
from tasks.llm.config import MAX_STEPS_LOWER_LEVEL, ANALYST_V2_MODEL
from tasks.llm.models import LlmSettings
from .agent import AnalystAgent
from .prompt_loader import get_prompt
from .tools import SearchDBSchema, EditFile, ExecuteQuery, CreateFile


ONBOARDING_MAX_STEPS = 15  # Keep it fast — schema scan + a few edits
DASHBOARD_MAX_STEPS = 25  # More steps needed: create questions + build dashboard


@register_agent
class OnboardingContextAgent(AnalystAgent):
    """
    Lightweight agent for the onboarding context step.

    Reads the database schema via SearchDBSchema and writes markdown
    documentation into the context file via EditFile. Deliberately
    limited in tools and prompt length to keep latency low.
    """

    def _get_system_message(self) -> dict:
        schema_str = json.dumps(self.schema, indent=2) if self.schema else "No schema provided."
        max_steps = min(ONBOARDING_MAX_STEPS, MAX_STEPS_LOWER_LEVEL - 5)
        content = get_prompt(
            'onboarding_context.system',
            agent_name=self.agent_name,
            schema=schema_str,
            connection_id=self.connection_id,
            max_steps=max_steps,
        )
        return {"role": "system", "content": content}

    def _get_user_message(self) -> dict:
        app_state_str = json.dumps(self.app_state, separators=(',', ':')) if self.app_state else "null"
        content = get_prompt(
            'onboarding_context.user',
            app_state=app_state_str,
            current_date=time.strftime("%Y-%m-%d"),
            goal=self.goal,
        )
        return {"role": "user", "content": content}

    def _get_llm_settings(self) -> LlmSettings:
        return LlmSettings(
            model=ANALYST_V2_MODEL,
            response_format={"type": "text"},
            tool_choice="auto",
            include_web_search=False,
        )

    def _get_available_tools(self):
        if len(self.tool_thread) >= ONBOARDING_MAX_STEPS:
            return []
        return [SearchDBSchema, EditFile, ExecuteQuery]


@register_agent
class OnboardingDashboardAgent(AnalystAgent):
    """
    Agent for the onboarding dashboard step.

    Creates 3-4 questions with varied visualizations and assembles them
    into a starter dashboard. Uses CreateFile for questions, EditFile for
    the dashboard layout, and ExecuteQuery/SearchDBSchema for exploration.
    """

    def _get_system_message(self) -> dict:
        schema_str = json.dumps(self.schema, indent=2) if self.schema else "No schema provided."
        max_steps = min(DASHBOARD_MAX_STEPS, MAX_STEPS_LOWER_LEVEL - 5)
        content = get_prompt(
            'onboarding_dashboard.system',
            agent_name=self.agent_name,
            schema=schema_str,
            context=self.context,
            connection_id=self.connection_id,
            max_steps=max_steps,
        )
        return {"role": "system", "content": content}

    def _get_user_message(self) -> dict:
        app_state_str = json.dumps(self.app_state, separators=(',', ':')) if self.app_state else "null"
        content = get_prompt(
            'onboarding_dashboard.user',
            app_state=app_state_str,
            current_date=time.strftime("%Y-%m-%d"),
            goal=self.goal,
        )
        return {"role": "user", "content": content}

    def _get_llm_settings(self) -> LlmSettings:
        return LlmSettings(
            model=ANALYST_V2_MODEL,
            response_format={"type": "text"},
            tool_choice="auto",
            include_web_search=False,
        )

    def _get_available_tools(self):
        if len(self.tool_thread) >= DASHBOARD_MAX_STEPS:
            return []
        return [SearchDBSchema, ExecuteQuery, CreateFile, EditFile]

File: frontend/agents/onboarding/onboarding-agents.ts
// Onboarding agents (OnboardingContextAgent / OnboardingDashboardAgent).
//
// Both run on the normal v=2 chat path (they use the frontend-bridged EditFile /
// CreateFile tools), so they extend WebAnalystAgent and just (a) restrict the
// toolset, (b) cap maxSteps for low latency, (c) render the onboarding-specific
// prompts (`onboarding_context.*` / `onboarding_dashboard.*`, already in
// `orchestrator/prompts/prompts.json`), and (d) disable web search
// (no web search).
import 'server-only';
import { Type } from 'typebox';
import type { TSchema } from 'typebox';
import type { Tool, TextContent, ImageContent } from '@/orchestrator/llm';
import { renderPrompt } from '@/orchestrator/prompts';
import { registerFauxProvider } from '@/orchestrator/llm/testing';
import { WebAnalystAgent, EditFile, CreateFile } from '@/agents/web-analyst/web-analyst';
import { SearchDBSchema, ExecuteQuery } from '@/agents/benchmark-analyst/db-tools.server';
import { getAnalystModel } from '@/agents/analyst/model-config';
import type { RemoteAnalystContext } from '@/agents/analyst/types';

export const fauxRegistration = registerFauxProvider({
  api: 'faux-onboarding-api',
  provider: 'faux-onboarding',
  models: [{ id: 'stub-onboarding' }],
});
const FAUX_MODEL = fauxRegistration.getModel();

const OnboardingAgentParams = Type.Object({
  userMessage: Type.String(),
});

function schemaString(ctx: RemoteAnalystContext): string {
  return ctx.schema ? JSON.stringify(ctx.schema, null, 2) : 'No schema provided.';
}

function goalFrom(raw: string | (TextContent | ImageContent)[]): string {
  if (typeof raw === 'string') return raw;
  return raw
    .filter((c): c is TextContent => c.type === 'text')
    .map((c) => c.text)
    .join('\n');
}

function onboardingUserContent(
  ctx: RemoteAnalystContext,
  raw: string | (TextContent | ImageContent)[],
  promptId: string,
): (TextContent | ImageContent)[] {
  const appState = ctx.appState !== undefined ? JSON.stringify(ctx.appState) : 'null';
  const text = renderPrompt(promptId, {
    app_state: appState,
    current_date: new Date().toISOString().slice(0, 10),
    goal: goalFrom(raw),
  });
  return [{ type: 'text', text }];
}

/**
 * Onboarding context step: reads schema (SearchDBSchema/ExecuteQuery) and writes
 * markdown docs into the context file via EditFile. Few tools, low max-steps.
 */
export class OnboardingContextAgent extends WebAnalystAgent {
  static override readonly schema: Tool<typeof OnboardingAgentParams> = {
    name: 'OnboardingContextAgent',
    description: 'Onboarding step that documents a database schema into a context file.',
    parameters: OnboardingAgentParams,
  };
  static override readonly tools: Tool<TSchema>[] = [
    SearchDBSchema.schema,
    EditFile.schema,
    ExecuteQuery.schema,
  ];
  static override readonly maxSteps = 15;
  static override model = getAnalystModel() ?? FAUX_MODEL;
  // No web search for onboarding.
  static override readonly callOptions = { reasoning: 'low', webSearch: false };

  protected override resolveCallOptions(): Record<string, unknown> | undefined {
    return (this.constructor as typeof OnboardingContextAgent).callOptions;
  }

  protected override getSystemPrompt(): string {
    const ctor = this.constructor as typeof OnboardingContextAgent;
    return renderPrompt('onboarding_context.system', {
      agent_name: this.context.agentName ?? 'MinusX',
      schema: schemaString(this.context),
      connection_id: this.context.connectionId ?? '',
      max_steps: String(ctor.maxSteps),
    });
  }

  protected override buildUserContent(): (TextContent | ImageContent)[] {
    return onboardingUserContent(this.context, this.userMessage, 'onboarding_context.user');
  }
}

/**
 * Onboarding dashboard step: creates a few questions (CreateFile) with varied
 * viz and assembles a starter dashboard (EditFile).
 */
export class OnboardingDashboardAgent extends WebAnalystAgent {
  static override readonly schema: Tool<typeof OnboardingAgentParams> = {
    name: 'OnboardingDashboardAgent',
    description: 'Onboarding step that creates starter questions and a dashboard.',
    parameters: OnboardingAgentParams,
  };
  static override readonly tools: Tool<TSchema>[] = [
    SearchDBSchema.schema,
    ExecuteQuery.schema,
    CreateFile.schema,
    EditFile.schema,
  ];
  static override readonly maxSteps = 25;
  static override model = getAnalystModel() ?? FAUX_MODEL;
  // No web search for onboarding.
  static override readonly callOptions = { reasoning: 'low', webSearch: false };

  protected override resolveCallOptions(): Record<string, unknown> | undefined {
    return (this.constructor as typeof OnboardingDashboardAgent).callOptions;
  }

  protected override getSystemPrompt(): string {
    const ctor = this.constructor as typeof OnboardingDashboardAgent;
    return renderPrompt('onboarding_dashboard.system', {
      agent_name: this.context.agentName ?? 'MinusX',
      schema: schemaString(this.context),
      context: this.context.contextDocs ?? '',
      connection_id: this.context.connectionId ?? '',
      max_steps: String(ctor.maxSteps),
    });
  }

  protected override buildUserContent(): (TextContent | ImageContent)[] {
    return onboardingUserContent(this.context, this.userMessage, 'onboarding_dashboard.user');
  }
}
